File: src/other-languages/java/tests/SanitizeByTierTest.java
// Copyright 2026 Daniel Smith
// Licensed under the Apache License, Version 2.0
// See https://www.apache.org/licenses/LICENSE-2.0

package tests;

import alap.LinkProvenance;
import alap.SanitizeByTier;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Tests for SanitizeByTier — tier-aware URL / cssClass / targetWindow.
 *
 * <p>Run: {@code java -ea tests.SanitizeByTierTest}
 */
public class SanitizeByTierTest {

    private static int passed = 0;
    private static int failed = 0;

    public static void main(String[] args) {
        System.out.println("Alap Java — SanitizeByTier Test Suite\n");

        section("URL — author tier");
        run(SanitizeByTierTest::testUrlAuthorKeepsHttps);
        run(SanitizeByTierTest::testUrlAuthorKeepsHttp);
        run(SanitizeByTierTest::testUrlAuthorKeepsTel);
        run(SanitizeByTierTest::testUrlAuthorKeepsMailto);
        run(SanitizeByTierTest::testUrlAuthorKeepsCustomScheme);
        run(SanitizeByTierTest::testUrlAuthorStillBlocksJavascript);
        run(SanitizeByTierTest::testUrlAuthorStillBlocksData);
        run(SanitizeByTierTest::testUrlAuthorKeepsRelative);

        section("URL — storage tier");
        run(SanitizeByTierTest::testUrlStorageRemoteKeepsHttps);
        run(SanitizeByTierTest::testUrlStorageRemoteKeepsMailto);
        run(SanitizeByTierTest::testUrlStorageRemoteRejectsTel);
        run(SanitizeByTierTest::testUrlStorageRemoteRejectsCustomScheme);
        run(SanitizeByTierTest::testUrlStorageLocalRejectsTel);
        run(SanitizeByTierTest::testUrlStorageRemoteStillBlocksJavascript);

        section("URL — protocol tier");
        run(SanitizeByTierTest::testUrlProtocolKeepsHttps);
        run(SanitizeByTierTest::testUrlProtocolRejectsTel);
        run(SanitizeByTierTest::testUrlProtocolRejectsCustomScheme);
        run(SanitizeByTierTest::testUrlProtocolBlocksJavascript);

        section("URL — unstamped (fail-closed)");
        run(SanitizeByTierTest::testUrlUnstampedRejectsTel);
        run(SanitizeByTierTest::testUrlUnstampedKeepsHttps);
        run(SanitizeByTierTest::testUrlUnstampedBlocksJavascript);

        section("cssClass");
        run(SanitizeByTierTest::testCssClassAuthorKeepsClass);
        run(SanitizeByTierTest::testCssClassAuthorKeepsMultiWord);
        run(SanitizeByTierTest::testCssClassAuthorNullStaysNull);
        run(SanitizeByTierTest::testCssClassStorageRemoteDropsClass);
        run(SanitizeByTierTest::testCssClassStorageLocalDropsClass);
        run(SanitizeByTierTest::testCssClassProtocolDropsClass);
        run(SanitizeByTierTest::testCssClassProtocolNullStaysNull);
        run(SanitizeByTierTest::testCssClassUnstampedDropsClass);

        section("targetWindow");
        run(SanitizeByTierTest::testTargetWindowAuthorKeepsSelf);
        run(SanitizeByTierTest::testTargetWindowAuthorKeepsBlank);
        run(SanitizeByTierTest::testTargetWindowAuthorKeepsNamedWindow);
        run(SanitizeByTierTest::testTargetWindowAuthorPassesNullThrough);
        run(SanitizeByTierTest::testTargetWindowStorageClampsSelfToBlank);
        run(SanitizeByTierTest::testTargetWindowStorageClampsNamedWindowToBlank);
        run(SanitizeByTierTest::testTargetWindowStorageClampsNullToBlank);
        run(SanitizeByTierTest::testTargetWindowStorageLocalClamps);
        run(SanitizeByTierTest::testTargetWindowProtocolClamps);
        run(SanitizeByTierTest::testTargetWindowUnstampedClamps);
        run(SanitizeByTierTest::testTargetWindowUnstampedNullClamps);

        System.out.println("\n==================================================");
        System.out.println(passed + " passed, " + failed + " failed, " + (passed + failed) + " total");
        System.exit(failed == 0 ? 0 : 1);
    }

    private static Map<String, Object> stampedLink(String tier) {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("url", "/a");
        LinkProvenance.stamp(m, tier);
        return m;
    }

    private static Map<String, Object> unstampedLink() {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("url", "/a");
        return m;
    }

    // --- URL: author ---

    static void testUrlAuthorKeepsHttps() {
        assertEquals("https://example.com",
            SanitizeByTier.url("https://example.com", stampedLink("author")),
            "author keeps https");
    }

    static void testUrlAuthorKeepsHttp() {
        assertEquals("http://example.com",
            SanitizeByTier.url("http://example.com", stampedLink("author")),
            "author keeps http");
    }

    static void testUrlAuthorKeepsTel() {
        assertEquals("tel:+15551234",
            SanitizeByTier.url("tel:+15551234", stampedLink("author")),
            "author keeps tel");
    }

    static void testUrlAuthorKeepsMailto() {
        assertEquals("mailto:a@b.com",
            SanitizeByTier.url("mailto:a@b.com", stampedLink("author")),
            "author keeps mailto");
    }

    static void testUrlAuthorKeepsCustomScheme() {
        assertEquals("obsidian://open?vault=foo",
            SanitizeByTier.url("obsidian://open?vault=foo", stampedLink("author")),
            "author keeps obsidian://");
    }

    static void testUrlAuthorStillBlocksJavascript() {
        assertEquals("about:blank",
            SanitizeByTier.url("javascript:alert(1)", stampedLink("author")),
            "author blocks javascript");
    }

    static void testUrlAuthorStillBlocksData() {
        assertEquals("about:blank",
            SanitizeByTier.url("data:text/html,x", stampedLink("author")),
            "author blocks data");
    }

    static void testUrlAuthorKeepsRelative() {
        assertEquals("/foo/bar",
            SanitizeByTier.url("/foo/bar", stampedLink("author")),
            "author keeps relative");
    }

    // --- URL: storage ---

    static void testUrlStorageRemoteKeepsHttps() {
        assertEquals("https://example.com",
            SanitizeByTier.url("https://example.com", stampedLink("storage:remote")),
            "storage:remote keeps https");
    }

    static void testUrlStorageRemoteKeepsMailto() {
        assertEquals("mailto:a@b.com",
            SanitizeByTier.url("mailto:a@b.com", stampedLink("storage:remote")),
            "storage:remote keeps mailto");
    }

    static void testUrlStorageRemoteRejectsTel() {
        assertEquals("about:blank",
            SanitizeByTier.url("tel:+15551234", stampedLink("storage:remote")),
            "storage:remote rejects tel");
    }

    static void testUrlStorageRemoteRejectsCustomScheme() {
        assertEquals("about:blank",
            SanitizeByTier.url("obsidian://open?vault=foo", stampedLink("storage:remote")),
            "storage:remote rejects custom scheme");
    }

    static void testUrlStorageLocalRejectsTel() {
        assertEquals("about:blank",
            SanitizeByTier.url("tel:+15551234", stampedLink("storage:local")),
            "storage:local rejects tel");
    }

    static void testUrlStorageRemoteStillBlocksJavascript() {
        assertEquals("about:blank",
            SanitizeByTier.url("javascript:alert(1)", stampedLink("storage:remote")),
            "storage:remote blocks javascript");
    }

    // --- URL: protocol ---

    static void testUrlProtocolKeepsHttps() {
        assertEquals("https://example.com",
            SanitizeByTier.url("https://example.com", stampedLink("protocol:web")),
            "protocol:web keeps https");
    }

    static void testUrlProtocolRejectsTel() {
        assertEquals("about:blank",
            SanitizeByTier.url("tel:+15551234", stampedLink("protocol:web")),
            "protocol:web rejects tel");
    }

    static void testUrlProtocolRejectsCustomScheme() {
        assertEquals("about:blank",
            SanitizeByTier.url("obsidian://open", stampedLink("protocol:atproto")),
            "protocol:atproto rejects custom scheme");
    }

    static void testUrlProtocolBlocksJavascript() {
        assertEquals("about:blank",
            SanitizeByTier.url("javascript:alert(1)", stampedLink("protocol:web")),
            "protocol:web blocks javascript");
    }

    // --- URL: unstamped (fail-closed) ---

    static void testUrlUnstampedRejectsTel() {
        assertEquals("about:blank",
            SanitizeByTier.url("tel:+15551234", unstampedLink()),
            "unstamped rejects tel");
    }

    static void testUrlUnstampedKeepsHttps() {
        assertEquals("https://example.com",
            SanitizeByTier.url("https://example.com", unstampedLink()),
            "unstamped keeps https");
    }

    static void testUrlUnstampedBlocksJavascript() {
        assertEquals("about:blank",
            SanitizeByTier.url("javascript:alert(1)", unstampedLink()),
            "unstamped blocks javascript");
    }

    // --- cssClass ---

    static void testCssClassAuthorKeepsClass() {
        assertEquals("my-class",
            SanitizeByTier.cssClass("my-class", stampedLink("author")),
            "author keeps class");
    }

    static void testCssClassAuthorKeepsMultiWord() {
        assertEquals("primary special",
            SanitizeByTier.cssClass("primary special", stampedLink("author")),
            "author keeps multi-word");
    }

    static void testCssClassAuthorNullStaysNull() {
        assertEquals(null,
            SanitizeByTier.cssClass(null, stampedLink("author")),
            "author null stays null");
    }

    static void testCssClassStorageRemoteDropsClass() {
        assertEquals(null,
            SanitizeByTier.cssClass("my-class", stampedLink("storage:remote")),
            "storage:remote drops class");
    }

    static void testCssClassStorageLocalDropsClass() {
        assertEquals(null,
            SanitizeByTier.cssClass("my-class", stampedLink("storage:local")),
            "storage:local drops class");
    }

    static void testCssClassProtocolDropsClass() {
        assertEquals(null,
            SanitizeByTier.cssClass("my-class", stampedLink("protocol:web")),
            "protocol drops class");
    }

    static void testCssClassProtocolNullStaysNull() {
        assertEquals(null,
            SanitizeByTier.cssClass(null, stampedLink("protocol:web")),
            "protocol null stays null");
    }

    static void testCssClassUnstampedDropsClass() {
        assertEquals(null,
            SanitizeByTier.cssClass("my-class", unstampedLink()),
            "unstamped drops class");
    }

    // --- targetWindow ---

    static void testTargetWindowAuthorKeepsSelf() {
        assertEquals("_self",
            SanitizeByTier.targetWindow("_self", stampedLink("author")),
            "author keeps _self");
    }

    static void testTargetWindowAuthorKeepsBlank() {
        assertEquals("_blank",
            SanitizeByTier.targetWindow("_blank", stampedLink("author")),
            "author keeps _blank");
    }

    static void testTargetWindowAuthorKeepsNamedWindow() {
        assertEquals("fromAlap",
            SanitizeByTier.targetWindow("fromAlap", stampedLink("author")),
            "author keeps named window");
    }

    static void testTargetWindowAuthorPassesNullThrough() {
        // Author-tier intentionally preserves null so the caller's
        // fallback chain still applies.
        assertEquals(null,
            SanitizeByTier.targetWindow(null, stampedLink("author")),
            "author passes null through");
    }

    static void testTargetWindowStorageClampsSelfToBlank() {
        assertEquals("_blank",
            SanitizeByTier.targetWindow("_self", stampedLink("storage:remote")),
            "storage:remote clamps _self");
    }

    static void testTargetWindowStorageClampsNamedWindowToBlank() {
        assertEquals("_blank",
            SanitizeByTier.targetWindow("fromAlap", stampedLink("storage:remote")),
            "storage:remote clamps named window");
    }

    static void testTargetWindowStorageClampsNullToBlank() {
        // Non-author tier forces _blank even when input is null.
        assertEquals("_blank",
            SanitizeByTier.targetWindow(null, stampedLink("storage:remote")),
            "storage:remote clamps null to _blank");
    }

    static void testTargetWindowStorageLocalClamps() {
        assertEquals("_blank",
            SanitizeByTier.targetWindow("_parent", stampedLink("storage:local")),
            "storage:local clamps _parent");
    }

    static void testTargetWindowProtocolClamps() {
        assertEquals("_blank",
            SanitizeByTier.targetWindow("fromAlap", stampedLink("protocol:web")),
            "protocol clamps named window");
    }

    static void testTargetWindowUnstampedClamps() {
        assertEquals("_blank",
            SanitizeByTier.targetWindow("_self", unstampedLink()),
            "unstamped clamps _self");
    }

    static void testTargetWindowUnstampedNullClamps() {
        assertEquals("_blank",
            SanitizeByTier.targetWindow(null, unstampedLink()),
            "unstamped null clamps to _blank");
    }

    // --- Runner ---

    private static void run(Runnable test) {
        try {
            test.run();
            passed++;
        } catch (AssertionError e) {
            failed++;
            System.out.println("  FAIL: " + e.getMessage());
        } catch (Throwable t) {
            failed++;
            System.out.println("  ERROR: " + t.getClass().getSimpleName() + ": " + t.getMessage());
        }
    }

    private static void section(String name) {
        System.out.println("\n--- " + name + " ---");
    }

    private static void assertEquals(Object expected, Object actual, String msg) {
        if (expected == null ? actual != null : !expected.equals(actual)) {
            throw new AssertionError(msg + ": expected " + expected + " but got " + actual);
        }
        System.out.println("  PASS: " + msg);
    }
}

File: src/other-languages/ruby/test/sanitize_by_tier_test.rb
# frozen_string_literal: true

# Copyright 2026 Daniel Smith
# Licensed under the Apache License, Version 2.0
# See https://www.apache.org/licenses/LICENSE-2.0

require "minitest/autorun"
require_relative "../lib/alap"

def stamped_link(tier)
  link = { "url" => "/a" }
  Alap::LinkProvenance.stamp(link, tier)
  link
end

# ---------------------------------------------------------------------------
# SanitizeByTier.url
# ---------------------------------------------------------------------------

class SanitizeByTierUrlAuthorTest < Minitest::Test
  def test_author_keeps_https
    assert_equal "https://example.com", Alap::SanitizeByTier.url("https://example.com", stamped_link("author"))
  end

  def test_author_keeps_http
    assert_equal "http://example.com", Alap::SanitizeByTier.url("http://example.com", stamped_link("author"))
  end

  def test_author_keeps_tel
    assert_equal "tel:+15551234", Alap::SanitizeByTier.url("tel:+15551234", stamped_link("author"))
  end

  def test_author_keeps_mailto
    assert_equal "mailto:a@b.com", Alap::SanitizeByTier.url("mailto:a@b.com", stamped_link("author"))
  end

  def test_author_keeps_custom_scheme
    assert_equal(
      "obsidian://open?vault=foo",
      Alap::SanitizeByTier.url("obsidian://open?vault=foo", stamped_link("author")),
    )
  end

  def test_author_still_blocks_javascript
    assert_equal "about:blank", Alap::SanitizeByTier.url("javascript:alert(1)", stamped_link("author"))
  end

  def test_author_still_blocks_data
    assert_equal "about:blank", Alap::SanitizeByTier.url("data:text/html,x", stamped_link("author"))
  end

  def test_author_keeps_relative
    assert_equal "/foo/bar", Alap::SanitizeByTier.url("/foo/bar", stamped_link("author"))
  end
end

class SanitizeByTierUrlStorageTest < Minitest::Test
  def test_storage_remote_keeps_https
    assert_equal(
      "https://example.com",
      Alap::SanitizeByTier.url("https://example.com", stamped_link("storage:remote")),
    )
  end

  def test_storage_remote_keeps_mailto
    assert_equal(
      "mailto:a@b.com",
      Alap::SanitizeByTier.url("mailto:a@b.com", stamped_link("storage:remote")),
    )
  end

  def test_storage_remote_rejects_tel
    assert_equal "about:blank", Alap::SanitizeByTier.url("tel:+15551234", stamped_link("storage:remote"))
  end

  def test_storage_remote_rejects_custom_scheme
    assert_equal(
      "about:blank",
      Alap::SanitizeByTier.url("obsidian://open?vault=foo", stamped_link("storage:remote")),
    )
  end

  def test_storage_local_rejects_tel
    assert_equal "about:blank", Alap::SanitizeByTier.url("tel:+15551234", stamped_link("storage:local"))
  end

  def test_storage_remote_still_blocks_javascript
    assert_equal(
      "about:blank",
      Alap::SanitizeByTier.url("javascript:alert(1)", stamped_link("storage:remote")),
    )
  end
end

class SanitizeByTierUrlProtocolTest < Minitest::Test
  def test_protocol_keeps_https
    assert_equal(
      "https://example.com",
      Alap::SanitizeByTier.url("https://example.com", stamped_link("protocol:web")),
    )
  end

  def test_protocol_rejects_tel
    assert_equal "about:blank", Alap::SanitizeByTier.url("tel:+15551234", stamped_link("protocol:web"))
  end

  def test_protocol_rejects_custom_scheme
    assert_equal "about:blank", Alap::SanitizeByTier.url("obsidian://open", stamped_link("protocol:atproto"))
  end

  def test_protocol_blocks_javascript
    assert_equal "about:blank", Alap::SanitizeByTier.url("javascript:alert(1)", stamped_link("protocol:web"))
  end
end

class SanitizeByTierUrlUnstampedTest < Minitest::Test
  def test_unstamped_rejects_tel
    link = { "url" => "/a" }
    assert_equal "about:blank", Alap::SanitizeByTier.url("tel:+15551234", link)
  end

  def test_unstamped_keeps_https
    link = { "url" => "/a" }
    assert_equal "https://example.com", Alap::SanitizeByTier.url("https://example.com", link)
  end

  def test_unstamped_blocks_javascript
    link = { "url" => "/a" }
    assert_equal "about:blank", Alap::SanitizeByTier.url("javascript:alert(1)", link)
  end
end

# ---------------------------------------------------------------------------
# SanitizeByTier.css_class
# ---------------------------------------------------------------------------

class SanitizeByTierCssClassTest < Minitest::Test
  def test_author_keeps_class
    assert_equal "my-class", Alap::SanitizeByTier.css_class("my-class", stamped_link("author"))
  end

  def test_author_keeps_multi_word
    assert_equal "primary special", Alap::SanitizeByTier.css_class("primary special", stamped_link("author"))
  end

  def test_author_nil_stays_nil
    assert_nil Alap::SanitizeByTier.css_class(nil, stamped_link("author"))
  end

  def test_storage_remote_drops_class
    assert_nil Alap::SanitizeByTier.css_class("my-class", stamped_link("storage:remote"))
  end

  def test_storage_local_drops_class
    assert_nil Alap::SanitizeByTier.css_class("my-class", stamped_link("storage:local"))
  end

  def test_protocol_drops_class
    assert_nil Alap::SanitizeByTier.css_class("my-class", stamped_link("protocol:web"))
  end

  def test_protocol_nil_stays_nil
    assert_nil Alap::SanitizeByTier.css_class(nil, stamped_link("protocol:web"))
  end

  def test_unstamped_drops_class
    link = { "url" => "/a" }
    assert_nil Alap::SanitizeByTier.css_class("my-class", link)
  end
end

# ---------------------------------------------------------------------------
# SanitizeByTier.target_window
# ---------------------------------------------------------------------------

class SanitizeByTierTargetWindowTest < Minitest::Test
  def test_author_keeps_self
    assert_equal "_self", Alap::SanitizeByTier.target_window("_self", stamped_link("author"))
  end

  def test_author_keeps_blank
    assert_equal "_blank", Alap::SanitizeByTier.target_window("_blank", stamped_link("author"))
  end

  def test_author_keeps_named_window
    assert_equal "fromAlap", Alap::SanitizeByTier.target_window("fromAlap", stamped_link("author"))
  end

  def test_author_passes_nil_through
    # Author-tier intentionally preserves nil so the caller's fallback
    # chain still applies.
    assert_nil Alap::SanitizeByTier.target_window(nil, stamped_link("author"))
  end

  def test_storage_clamps_self_to_blank
    assert_equal "_blank", Alap::SanitizeByTier.target_window("_self", stamped_link("storage:remote"))
  end

  def test_storage_clamps_named_window_to_blank
    assert_equal "_blank", Alap::SanitizeByTier.target_window("fromAlap", stamped_link("storage:remote"))
  end

  def test_storage_clamps_nil_to_blank
    # Non-author tier forces _blank even when input is nil, so a missing
    # targetWindow does not inherit author-tier defaults.
    assert_equal "_blank", Alap::SanitizeByTier.target_window(nil, stamped_link("storage:remote"))
  end

  def test_storage_local_clamps
    assert_equal "_blank", Alap::SanitizeByTier.target_window("_parent", stamped_link("storage:local"))
  end

  def test_protocol_clamps
    assert_equal "_blank", Alap::SanitizeByTier.target_window("fromAlap", stamped_link("protocol:web"))
  end

  def test_unstamped_clamps
    link = { "url" => "/a" }
    assert_equal "_blank", Alap::SanitizeByTier.target_window("_self", link)
  end

  def test_unstamped_nil_clamps
    link = { "url" => "/a" }
    assert_equal "_blank", Alap::SanitizeByTier.target_window(nil, link)
  end
end
